### PR TITLE
Corr config

### PR DESCRIPTION
**Pull request by david-deboer**
This puts the info in the correlator configurations into a set of M&C tables. The configurations were already being tracked via hashes linked to files in the librarian, but this gets the info out into tables so it can be queried directly.

Still to do:
- ~update the documentation to include these new tables.~

### CODE REVIEW

**Comment by david-deboer** [comment]
For instance, here is the current config:

         config_file_hash         |             parameter              |                         value 
 aff3a9129a3716293cd379f971d7f75a | fft_shift                          | 15086
 aff3a9129a3716293cd379f971d7f75a | fpgfile                            | redis:snap_fengine_2020-07-16_1253.fpg
 aff3a9129a3716293cd379f971d7f75a | dest_port                          | 8511
 aff3a9129a3716293cd379f971d7f75a | log_walsh_step_size                | 3
 aff3a9129a3716293cd379f971d7f75a | walsh_order                        | 32
 aff3a9129a3716293cd379f971d7f75a | walsh_delay                        | 600
 aff3a9129a3716293cd379f971d7f75a | fengines                           | heraNode0Snap0,heraNode0Snap1,heraNode0Snap2,heraNode0Snap3,heraNode10Snap0,heraNode10Snap1,heraNode10Snap2,heraNode10Snap3,heraNode12Snap0,heraNode12Snap1,heraNode12Snap2,heraNode12Snap3,heraNode13Snap0,heraNode13Snap1,heraNode13Snap2,heraNode13Snap3,heraNode14Snap0,heraNode14Snap1,heraNode14Snap2,heraNode14Snap3,heraNode3Snap0,heraNode3Snap1,heraNode3Snap2,heraNode3Snap3,heraNode5Snap0,heraNode5Snap1,heraNode5Snap3,heraNode7Snap0,heraNode7Snap1,heraNode7Snap2,heraNode7Snap3,heraNode8Snap0,heraNode8Snap1,heraNode8Snap2,heraNode8Snap3,heraNode9Snap0,heraNode9Snap1,heraNode9Snap2
 aff3a9129a3716293cd379f971d7f75a | heraNode0Snap0:ants                | {0,1,2}
 aff3a9129a3716293cd379f971d7f75a | heraNode0Snap0:phase_switch_index  | {1,2,3,4,5,6}
 aff3a9129a3716293cd379f971d7f75a | heraNode0Snap1:ants                | {3,4,5}
 aff3a9129a3716293cd379f971d7f75a | heraNode0Snap1:phase_switch_index  | {7,8,9,10,11,12}
 aff3a9129a3716293cd379f971d7f75a | heraNode0Snap2:ants                | {6,7,8}
 aff3a9129a3716293cd379f971d7f75a | heraNode0Snap2:phase_switch_index  | {13,14,15,16,17,18}
 aff3a9129a3716293cd379f971d7f75a | heraNode0Snap3:ants                | {9,10,11}
 aff3a9129a3716293cd379f971d7f75a | heraNode0Snap3:phase_switch_index  | {19,20,21,22,23,24}
 aff3a9129a3716293cd379f971d7f75a | heraNode10Snap0:ants               | {108,109,110}
 aff3a9129a3716293cd379f971d7f75a | heraNode10Snap0:phase_switch_index | {1,2,3,4,5,6}
 aff3a9129a3716293cd379f971d7f75a | heraNode10Snap1:ants               | {111,112,113}
 aff3a9129a3716293cd379f971d7f75a | heraNode10Snap1:phase_switch_index | {7,8,9,10,11,12}
 aff3a9129a3716293cd379f971d7f75a | heraNode10Snap2:ants               | {114,115,116}
 aff3a9129a3716293cd379f971d7f75a | heraNode10Snap2:phase_switch_index | {13,14,15,16,17,18}
 aff3a9129a3716293cd379f971d7f75a | heraNode10Snap3:ants               | {117,118,119}
 aff3a9129a3716293cd379f971d7f75a | heraNode10Snap3:phase_switch_index | {19,20,21,22,23,24}
 aff3a9129a3716293cd379f971d7f75a | heraNode12Snap0:ants               | {84,85,86}
 aff3a9129a3716293cd379f971d7f75a | heraNode12Snap0:phase_switch_index | {1,2,3,4,5,6}
 aff3a9129a3716293cd379f971d7f75a | heraNode12Snap1:ants               | {87,88,89}
 aff3a9129a3716293cd379f971d7f75a | heraNode12Snap1:phase_switch_index | {7,8,9,10,11,12}
 aff3a9129a3716293cd379f971d7f75a | heraNode12Snap2:ants               | {90,91,92}
 aff3a9129a3716293cd379f971d7f75a | heraNode12Snap2:phase_switch_index | {13,14,15,16,17,18}
 aff3a9129a3716293cd379f971d7f75a | heraNode12Snap3:ants               | {93,94,95}
 aff3a9129a3716293cd379f971d7f75a | heraNode12Snap3:phase_switch_index | {19,20,21,22,23,24}
 aff3a9129a3716293cd379f971d7f75a | heraNode13Snap0:ants               | {72,73,74}
 aff3a9129a3716293cd379f971d7f75a | heraNode13Snap0:phase_switch_index | {1,2,3,4,5,6}
 aff3a9129a3716293cd379f971d7f75a | heraNode13Snap1:ants               | {75,76,77}
 aff3a9129a3716293cd379f971d7f75a | heraNode13Snap1:phase_switch_index | {7,8,9,10,11,12}
 aff3a9129a3716293cd379f971d7f75a | heraNode13Snap2:ants               | {78,79,80}
 aff3a9129a3716293cd379f971d7f75a | heraNode13Snap2:phase_switch_index | {13,14,15,16,17,18}
 aff3a9129a3716293cd379f971d7f75a | heraNode13Snap3:ants               | {81,82,83}
 aff3a9129a3716293cd379f971d7f75a | heraNode13Snap3:phase_switch_index | {19,20,21,22,23,24}
 aff3a9129a3716293cd379f971d7f75a | heraNode14Snap0:ants               | {120,121,122}
 aff3a9129a3716293cd379f971d7f75a | heraNode14Snap0:phase_switch_index | {1,2,3,4,5,6}
 aff3a9129a3716293cd379f971d7f75a | heraNode14Snap1:ants               | {123,124,125}
 aff3a9129a3716293cd379f971d7f75a | heraNode14Snap1:phase_switch_index | {7,8,9,10,11,12}
 aff3a9129a3716293cd379f971d7f75a | heraNode14Snap2:ants               | {126,127,128}
 aff3a9129a3716293cd379f971d7f75a | heraNode14Snap2:phase_switch_index | {13,14,15,16,17,18}
 aff3a9129a3716293cd379f971d7f75a | heraNode14Snap3:ants               | {129,130,131}
 aff3a9129a3716293cd379f971d7f75a | heraNode14Snap3:phase_switch_index | {19,20,21,22,23,24}
 aff3a9129a3716293cd379f971d7f75a | heraNode3Snap0:ants                | {96,97,98}
 aff3a9129a3716293cd379f971d7f75a | heraNode3Snap0:phase_switch_index  | {1,2,3,4,5,6}
 aff3a9129a3716293cd379f971d7f75a | heraNode3Snap1:ants                | {99,100,101}
 aff3a9129a3716293cd379f971d7f75a | heraNode3Snap1:phase_switch_index  | {7,8,9,10,11,12}
 aff3a9129a3716293cd379f971d7f75a | heraNode3Snap2:ants                | {102,103,104}
 aff3a9129a3716293cd379f971d7f75a | heraNode3Snap2:phase_switch_index  | {13,14,15,16,17,18}
 aff3a9129a3716293cd379f971d7f75a | heraNode3Snap3:ants                | {105,106,107}
 aff3a9129a3716293cd379f971d7f75a | heraNode3Snap3:phase_switch_index  | {19,20,21,22,23,24}
 aff3a9129a3716293cd379f971d7f75a | heraNode5Snap0:ants                | {60,61,62}
 aff3a9129a3716293cd379f971d7f75a | heraNode5Snap0:phase_switch_index  | {1,2,3,4,5,6}
 aff3a9129a3716293cd379f971d7f75a | heraNode5Snap1:ants                | {63,64,65}
 aff3a9129a3716293cd379f971d7f75a | heraNode5Snap1:phase_switch_index  | {7,8,9,10,11,12}
 aff3a9129a3716293cd379f971d7f75a | heraNode5Snap3:ants                | {69,70,71}
 aff3a9129a3716293cd379f971d7f75a | heraNode5Snap3:phase_switch_index  | {19,20,21,22,23,24}
 aff3a9129a3716293cd379f971d7f75a | heraNode7Snap0:ants                | {36,37,38}
 aff3a9129a3716293cd379f971d7f75a | heraNode7Snap0:phase_switch_index  | {1,2,3,4,5,6}
 aff3a9129a3716293cd379f971d7f75a | heraNode7Snap1:ants                | {39,40,41}
 aff3a9129a3716293cd379f971d7f75a | heraNode7Snap1:phase_switch_index  | {7,8,9,10,11,12}
 aff3a9129a3716293cd379f971d7f75a | heraNode7Snap2:ants                | {42,43,44}
 aff3a9129a3716293cd379f971d7f75a | heraNode7Snap2:phase_switch_index  | {13,14,15,16,17,18}
 aff3a9129a3716293cd379f971d7f75a | heraNode7Snap3:ants                | {45,46,47}
 aff3a9129a3716293cd379f971d7f75a | heraNode7Snap3:phase_switch_index  | {19,20,21,22,23,24}
 aff3a9129a3716293cd379f971d7f75a | heraNode8Snap0:ants                | {12,13,14}
 aff3a9129a3716293cd379f971d7f75a | heraNode8Snap0:phase_switch_index  | {1,2,3,4,5,6}
 aff3a9129a3716293cd379f971d7f75a | heraNode8Snap1:ants                | {15,16,17}
 aff3a9129a3716293cd379f971d7f75a | heraNode8Snap1:phase_switch_index  | {7,8,9,10,11,12}
 aff3a9129a3716293cd379f971d7f75a | heraNode8Snap2:ants                | {18,19,20}
 aff3a9129a3716293cd379f971d7f75a | heraNode8Snap2:phase_switch_index  | {13,14,15,16,17,18}
 aff3a9129a3716293cd379f971d7f75a | heraNode8Snap3:ants                | {21,22,23}
 aff3a9129a3716293cd379f971d7f75a | heraNode8Snap3:phase_switch_index  | {19,20,21,22,23,24}
 aff3a9129a3716293cd379f971d7f75a | heraNode9Snap0:ants                | {24,25,26}
 aff3a9129a3716293cd379f971d7f75a | heraNode9Snap0:phase_switch_index  | {1,2,3,4,5,6}
 aff3a9129a3716293cd379f971d7f75a | heraNode9Snap1:ants                | {27,28,29}
 aff3a9129a3716293cd379f971d7f75a | heraNode9Snap1:phase_switch_index  | {7,8,9,10,11,12}
 aff3a9129a3716293cd379f971d7f75a | heraNode9Snap2:ants                | {30,31,32}
 aff3a9129a3716293cd379f971d7f75a | heraNode9Snap2:phase_switch_index  | {13,14,15,16,17,18}
 aff3a9129a3716293cd379f971d7f75a | xengines                           | 0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15
 aff3a9129a3716293cd379f971d7f75a | 0:chan_range                       | 1536,1920
 aff3a9129a3716293cd379f971d7f75a | 0:even:ip                          | 10.80.40.197
 aff3a9129a3716293cd379f971d7f75a | 0:even:mac                         | 2207786215621
 aff3a9129a3716293cd379f971d7f75a | 0:odd:ip                           | 10.80.40.206
 aff3a9129a3716293cd379f971d7f75a | 0:odd:mac                          | 2207786215630
 aff3a9129a3716293cd379f971d7f75a | 1:chan_range                       | 1920,2304
 aff3a9129a3716293cd379f971d7f75a | 1:even:ip                          | 10.80.40.229
 aff3a9129a3716293cd379f971d7f75a | 1:even:mac                         | 2207786215653
 aff3a9129a3716293cd379f971d7f75a | 1:odd:ip                           | 10.80.40.238
 aff3a9129a3716293cd379f971d7f75a | 1:odd:mac                          | 2207786215662
 aff3a9129a3716293cd379f971d7f75a | 2:chan_range                       | 2304,2688
 aff3a9129a3716293cd379f971d7f75a | 2:even:ip                          | 10.80.40.198
 aff3a9129a3716293cd379f971d7f75a | 2:even:mac                         | 2207786215622
 aff3a9129a3716293cd379f971d7f75a | 2:odd:ip                           | 10.80.40.207
 aff3a9129a3716293cd379f971d7f75a | 2:odd:mac                          | 2207786215631
 aff3a9129a3716293cd379f971d7f75a | 3:chan_range                       | 2688,3072
 aff3a9129a3716293cd379f971d7f75a | 3:even:ip                          | 10.80.40.230
 aff3a9129a3716293cd379f971d7f75a | 3:even:mac                         | 2207786215654
 aff3a9129a3716293cd379f971d7f75a | 3:odd:ip                           | 10.80.40.239
 aff3a9129a3716293cd379f971d7f75a | 3:odd:mac                          | 2207786215663
 aff3a9129a3716293cd379f971d7f75a | 4:chan_range                       | 3072,3456
 aff3a9129a3716293cd379f971d7f75a | 4:even:ip                          | 10.80.40.199
 aff3a9129a3716293cd379f971d7f75a | 4:even:mac                         | 2207786215623
 aff3a9129a3716293cd379f971d7f75a | 4:odd:ip                           | 10.80.40.208
 aff3a9129a3716293cd379f971d7f75a | 4:odd:mac                          | 2207786215632
 aff3a9129a3716293cd379f971d7f75a | 5:chan_range                       | 3456,3840
 aff3a9129a3716293cd379f971d7f75a | 5:even:ip                          | 10.80.40.231
 aff3a9129a3716293cd379f971d7f75a | 5:even:mac                         | 2207786215655
 aff3a9129a3716293cd379f971d7f75a | 5:odd:ip                           | 10.80.40.240
 aff3a9129a3716293cd379f971d7f75a | 5:odd:mac                          | 2207786215664
 aff3a9129a3716293cd379f971d7f75a | 6:chan_range                       | 3840,4224
 aff3a9129a3716293cd379f971d7f75a | 6:even:ip                          | 10.80.40.200
 aff3a9129a3716293cd379f971d7f75a | 6:even:mac                         | 2207786215624
 aff3a9129a3716293cd379f971d7f75a | 6:odd:ip                           | 10.80.40.209
 aff3a9129a3716293cd379f971d7f75a | 6:odd:mac                          | 2207786215633
 aff3a9129a3716293cd379f971d7f75a | 7:chan_range                       | 4224,4608
 aff3a9129a3716293cd379f971d7f75a | 7:even:ip                          | 10.80.40.232
 aff3a9129a3716293cd379f971d7f75a | 7:even:mac                         | 2207786215656
 aff3a9129a3716293cd379f971d7f75a | 7:odd:ip                           | 10.80.40.241
 aff3a9129a3716293cd379f971d7f75a | 7:odd:mac                          | 2207786215665
 aff3a9129a3716293cd379f971d7f75a | 8:chan_range                       | 4608,4992
 aff3a9129a3716293cd379f971d7f75a | 8:even:ip                          | 10.80.40.201
 aff3a9129a3716293cd379f971d7f75a | 8:even:mac                         | 2207786215625
 aff3a9129a3716293cd379f971d7f75a | 8:odd:ip                           | 10.80.40.210
 aff3a9129a3716293cd379f971d7f75a | 8:odd:mac                          | 2207786215634
 aff3a9129a3716293cd379f971d7f75a | 9:chan_range                       | 4992,5376
 aff3a9129a3716293cd379f971d7f75a | 9:even:ip                          | 10.80.40.233
 aff3a9129a3716293cd379f971d7f75a | 9:even:mac                         | 2207786215657
 aff3a9129a3716293cd379f971d7f75a | 9:odd:ip                           | 10.80.40.242
 aff3a9129a3716293cd379f971d7f75a | 9:odd:mac                          | 2207786215666
 aff3a9129a3716293cd379f971d7f75a | 10:chan_range                      | 5376,5760
 aff3a9129a3716293cd379f971d7f75a | 10:even:ip                         | 10.80.40.202
 aff3a9129a3716293cd379f971d7f75a | 10:even:mac                        | 2207786215626
 aff3a9129a3716293cd379f971d7f75a | 10:odd:ip                          | 10.80.40.211
 aff3a9129a3716293cd379f971d7f75a | 10:odd:mac                         | 2207786215635
 aff3a9129a3716293cd379f971d7f75a | 11:chan_range                      | 5760,6144
 aff3a9129a3716293cd379f971d7f75a | 11:even:ip                         | 10.80.40.234
 aff3a9129a3716293cd379f971d7f75a | 11:even:mac                        | 2207786215658
 aff3a9129a3716293cd379f971d7f75a | 11:odd:ip                          | 10.80.40.243
 aff3a9129a3716293cd379f971d7f75a | 11:odd:mac                         | 2207786215667
 aff3a9129a3716293cd379f971d7f75a | 12:chan_range                      | 6144,6528
 aff3a9129a3716293cd379f971d7f75a | 12:even:ip                         | 10.80.40.203
 aff3a9129a3716293cd379f971d7f75a | 12:even:mac                        | 2207786215627
 aff3a9129a3716293cd379f971d7f75a | 12:odd:ip                          | 10.80.40.212
 aff3a9129a3716293cd379f971d7f75a | 12:odd:mac                         | 2207786215636
 aff3a9129a3716293cd379f971d7f75a | 13:chan_range                      | 6528,6912
 aff3a9129a3716293cd379f971d7f75a | 13:even:ip                         | 10.80.40.235
 aff3a9129a3716293cd379f971d7f75a | 13:even:mac                        | 2207786215659
 aff3a9129a3716293cd379f971d7f75a | 13:odd:ip                          | 10.80.40.244
 aff3a9129a3716293cd379f971d7f75a | 13:odd:mac                         | 2207786215668
 aff3a9129a3716293cd379f971d7f75a | 14:chan_range                      | 6912,7296
 aff3a9129a3716293cd379f971d7f75a | 14:even:ip                         | 10.80.40.204
 aff3a9129a3716293cd379f971d7f75a | 14:even:mac                        | 2207786215628
 aff3a9129a3716293cd379f971d7f75a | 14:odd:ip                          | 10.80.40.213
 aff3a9129a3716293cd379f971d7f75a | 14:odd:mac                         | 2207786215637
 aff3a9129a3716293cd379f971d7f75a | 15:chan_range                      | 7296,7680
 aff3a9129a3716293cd379f971d7f75a | 15:even:ip                         | 10.80.40.236
 aff3a9129a3716293cd379f971d7f75a | 15:even:mac                        | 2207786215660
 aff3a9129a3716293cd379f971d7f75a | 15:odd:ip                          | 10.80.40.245
 aff3a9129a3716293cd379f971d7f75a | 15:odd:mac                         | 2207786215669

**Comment by bhazelton** [comment]
Let's split some of the info out into their own tables for better DB searchability:

corr_active_snap (multiple rows per hash)
- hash*
- node
- snap
- hostname*

corr_input_index (3 rows per hash/node/snap)
- hash*
- node
- snap
- hostname*
- corr_index
- ant_snap_position

corr_phase_switch_index
- hash*
- node
- snap
- hostname*
- phase_switch_index
- antpol_snap_position

**Comment by david-deboer** [comment]
This now makes 4 new tables, as outlined above.  Summarizing the 3 add'l from before:
correlator_active_snap:  config_file_hash*, hostname*, node, snap_position
correlator_input_index: config_file_hash*, hostname, node, snap_position, antenna_index_position, correlator_index*
correlator_phase_switch_index: config_file_hash*, hostname*, node, snap_position, antpol_index_position, phase_switch_index*

File: hera_mc/cm_redis_corr.py [hunk 1]
```diff
@@ -7,13 +7,102 @@
...
+    rsession = redis.Redis(connection_pool=redis_pool)
+
+    snap_configuration = 'snap_configuration'
+    if testing:
+        snap_configuration = 'testing_snap_configuration'
+
+    config = yaml.safe_load(rsession.hget(snap_configuration, 'config'))
+    md5 = rsession.hget(snap_configuration, 'md5').decode()
+
+    #####
```
Comment (bhazelton) [inline hera_mc/cm_redis_corr.py:46]: I think this comment line can come out.

File: alembic/versions/9e208dfd0778_add_correlator_config_table.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+# revision identifiers, used by Alembic.
+revision = '9e208dfd0778'
+down_revision = '7463268309ab'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (bhazelton) [inline alembic/versions/9e208dfd0778_add_correlator_config_table.py:20]: This and similar lines should be removed -- I'm sure that you made changes beyond the auto generated commands.

File: alembic/versions/c9e558e762ff_add_more_corr_config_tables.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+# revision identifiers, used by Alembic.
+revision = 'c9e558e762ff'
+down_revision = '9e208dfd0778'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (bhazelton) [inline alembic/versions/c9e558e762ff_add_more_corr_config_tables.py:20]: same comment in this file.

File: hera_mc/cm_redis_corr.py [hunk 4]
```diff
@@ -7,13 +7,102 @@
...
+                                                                     node=this_node,
+                                                                     snap_loc_num=this_snap,
+                                                                     antpol_index_position=_i,
+                                                                     phase_switch_index=pind))
+    xengines = [str(x) for x in sorted(config['xengines'].keys())]
+    session.add(correlator.CorrelatorConfiguration.create(config_file_hash=md5,
+                                                          parameter='xengines',
+                                                          value=','.join(xengines)))
+    for xengind, xeng in config['xengines'].items():
+        parameter = 'x{}:chan_range'.format(xengind)
```
Comment (bhazelton) [inline hera_mc/cm_redis_corr.py:91]: you could use an fstring here...

File: hera_mc/cm_redis_corr.py [hunk 5]
```diff
@@ -7,13 +7,102 @@
...
+        value = ','.join([str(_x) for _x in xeng['chan_range']])
+        session.add(correlator.CorrelatorConfiguration.create(config_file_hash=md5,
+                                                              parameter=parameter,
+                                                              value=value))
+        for evod in ['even', 'odd']:
+            for ipma in ['ip', 'mac']:
+                parameter = 'x{}:{}:{}'.format(xengind, evod, ipma)
```
Comment (bhazelton) [inline hera_mc/cm_redis_corr.py:98]: fstrings are awesome...

**Review by bhazelton — COMMENTED** [review]
I have one major comment: This is breaking with the established real time (as opposed to config management) structure, where we put the methods that require a session in the `mc_session.py` file as a method on the MCSession object and then create appropriate `add` and `get` methods there as well. We have a very standard setup for all the other tables in the `correlator.py` module (as well as other real time tables) that I really think we should follow, rather than putting this real time functionality into an amalgam of real time and config management modules. And then to ensure that this table is being filled out consistently, we can add the call to those methods to the existing `mc_monitor_correlator.py` script. If you do not feel comfortable following the established real time structure, I can help with that.

**Comment by david-deboer** [comment]
I would be happy for you to give a go.  I'm assuming that you just move/rename the code in cm_redis_corr.parse_snap_config_to_psql to mc_session and get rid of the session explicit stuff...

File: hera_mc/cm_redis_corr.py (hunk 5 above)
Comment (david-deboer) [reply hera_mc/cm_redis_corr.py]: They are cool

**Comment by bhazelton** [comment]
Yes, and make a couple new methods to have the standard set of methods for these types of tables. I'll take a hack at it soon.

**Comment by bhazelton** [comment]
@david-deboer I don't understand why some lines in `cm_sysutils` lost coverage. Can you take a look? Details here: https://codecov.io/gh/HERA-Team/hera_mc/pull/518/changes

**Comment by bhazelton** [comment]
@plaplant Are you willing to review this PR? Both Dave and I contributed to it significantly.

**Comment by plaplant** [comment]
@bhazelton sure, I will try to get to it in the next few days.

File: hera_mc/correlator.py [hunk 6]
```diff
@@ -268,6 +268,284 @@ def _get_config(corr_cm=None, correlator_redis_address=DEFAULT_REDIS_ADDRESS):
...
+            obj_list.append(
+                CorrelatorConfigInputIndex.create(
+                    config_hash=config_hash,
+                    correlator_index=cind,
+                    hostname=hostname,
+                    antenna_index_position=antind,
+                )
+            )
+        for psind, pind in enumerate(entry['phase_switch_index']):
+            pind = entry['phase_switch_index'][psind]
```
Comment (plaplant) [inline hera_mc/correlator.py:536]: If you're going to immediately overwrite the value of `pind`, it's better to make it just an underscore (`_`) in the `enumerate` construction. So something like:
```
    for psind, _ in enumerate(entry["phase_switch_index"]):
        pind = entry["phase_switch_index"][psind]
```
At that point, though, it seems like you really only need an index, so you could do something like:
```
    for psind in range(len(entry["phase_switch_index"]):
        pind = entry["phase_switch_index"][psind]
```

File: hera_mc/mc_session.py [hunk 7]
```diff
@@ -2307,26 +2307,588 @@ def add_correlator_config_file(self, config_hash, config_file):
...
         -------
-        list of CorrelatorConfigFile objects
+        list
+            List of CorrelatorConfigFile objects.
+        time_list : list of Astropy Time objects, optional
+            List of start times when the config files were used, same length as the list
+            of CorrelatorConfigFile objects. Only provided if `config_hash` is None.
 
         """
         from .correlator import CorrelatorConfigFile
```
Comment (plaplant) [inline hera_mc/mc_session.py:2350]: Is there a benefit to burying the import here instead of doing it at the top of the file? I generally prefer to do all imports at the top, unless there's a good reason not to (like it takes a long time or relies on optional modules).

File: hera_mc/mc_session.py [hunk 8]
```diff
@@ -2470,17 +3037,22 @@ def add_correlator_config_from_corrcm(self, config_state_dict=None,
         same_config_status = self.get_correlator_config_status(
             most_recent=True, config_hash=config_hash)
         if len(same_config_status) == 0:
+            # There's not a status with this config hash.
```
Comment (plaplant) [inline hera_mc/mc_session.py:3040]: Thanks for the informative comments 👍

**Review by plaplant — COMMENTED** [review]
Thanks for all the changes @david-deboer and @bhazelton! I have a few small comments, mostly about style rather than the main functionality.

File: hera_mc/cm_partconnect.py [hunk 9]
```diff
@@ -986,7 +986,7 @@ def update_connection(session=None, data=None, add_new_connection=False):
                 connection = None
             else:
                 connection = conn_rec.first()
-        else:
+        else:  # pragma: no cover
             print("Error:  more than one of ", dkey, " exists (which should not happen).")
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply hera_mc/cm_partconnect.py:990]: I agree. This is @david-deboer's part of the code, so I'll let him respond.

File: scripts/mc_parse_snap_config_to_psql.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
+#! /usr/bin/env python
+# -*- mode: python; coding: utf-8 -*-
+# Copyright 2020 the HERA Collaboration
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply scripts/mc_parse_snap_config_to_psql.py]: Yes! But I actually deleted this script because it points to code I removed.

File: hera_mc/tests/test_sysutils.py [hunk 11]
```diff
@@ -75,7 +75,7 @@ def test_watch_dog(mcsession):
 def test_ever_fully_connected(sys_handle):
     now_list = sys_handle.get_connected_stations(
         at_date='now', hookup_type='parts_hera')
-    assert len(now_list) == 12
+    assert len(now_list) == 11
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply hera_mc/tests/test_sysutils.py:78]: I did not understand why these changed when I made my very small tweaks to the test data. @david-deboer ?

File: hera_mc/data/test_data/initialization_data_part_rosetta.csv [hunk 12]
```diff
@@ -2,3 +2,7 @@ hpn,syspn,start_gpstime,stop_gptime
 SNPC000700,heraNode700Snap700,1275040818,
 SNPC000701,heraNode0Snap701,1275040818,
 SNPC000709,heraNode700Snap709,1275040818,
+SNPA000700,heraNode700Snap0,1230372018,
+SNPB000701,heraNode700Snap1,1230372018
+SNPC000702,heraNode700Snap2,1230372018
+SNPD007003,heraNode700Snap3,1230372018
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply hera_mc/data/test_data/initialization_data_part_rosetta.csv:8]: Yes. I'll add the line. I have been waiting for a break in PR traffic on this repo to make a push towards using black and pre-commit.

File: Docs/mc_definition.tex [hunk 13]
```diff
@@ -453,14 +453,63 @@ \subsubsection{correlator\_config\_file}
...
+\hline
+ {\bf Column} & {\bf Type}  & {\bf Description} \\ [0.5ex]  \hline\hline
+\textbf{\textit{config\_hash}} & string & hash for the config in use, foreign key into correlator\_config\_file table.\\ \hline
+\textbf{correlator\_index} & integer & Correlator index value (in the range 0 - 349). \\ \hline
+hostname* & string & Hostname of the SNAP (typically e.g. heraNode1Snap2). \\ \hline
+antenna\_index\_position* & integer & Antenna index position within the SNAP (in the range 0 - 2). \\ \hline
+\end{tabular}
+\end{center}
+
+\subsubsection{correlator\_config\_input\_index}
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply Docs/mc_definition.tex:492]: No, that's  a mistake, thanks!

File: hera_mc/mc_session.py (hunk 7 above)
Comment (bhazelton) [reply hera_mc/mc_session.py:2350]: I don't think there is a real reason to do it in the method, but it is the pattern in all the other methods in this object. I'll play around with moving them up to the top.

File: hera_mc/cm_partconnect.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (david-deboer) [reply hera_mc/cm_partconnect.py:990]: It would be an error.  It shouldn't be able to happen, given unique primary keys, but I left it in just in case reality is more clever than me in making it happen.  That is why the pragma -- I couldn't come up with a way to test it.  Bryna, since you are editing, can you raise a runtime error?  The other option is to remove since it shouldn't be able to happen, but then it would be silent.

File: hera_mc/cm_partconnect.py [hunk 14]
```diff
@@ -986,7 +986,7 @@ def update_connection(session=None, data=None, add_new_connection=False):
                 connection = None
             else:
                 connection = conn_rec.first()
-        else:
+        else:  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (david-deboer) [reply hera_mc/cm_partconnect.py:989]: See above

File: hera_mc/tests/test_sysutils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (david-deboer) [reply hera_mc/tests/test_sysutils.py:78]: These handle edge cases, so any tweaks to the test data have a chance to make them fail.  When I changed the number I verified it was correct.

File: hera_mc/mc_session.py (hunk 7 above)
Comment (bhazelton) [reply hera_mc/mc_session.py]: I moved all the imports from this package and from standard lib or our requirements to the top.

**Comment by bhazelton** [comment]
@plaplant I think I addressed all your comments. In the process of fixing the imports in `mc_session.py` I discovered that the way the data folder was being identified throughout the repo was strange, so I also fixed that (and that prompted me to look at the manifest where I found some out of date items related to versioning before we adopted setuptools_scm that I removed).

File: hera_mc/data/test_data/initialization_data_part_rosetta.csv [hunk 15]
```diff
@@ -2,3 +2,7 @@ hpn,syspn,start_gpstime,stop_gptime
 SNPC000700,heraNode700Snap700,1275040818,
 SNPC000701,heraNode0Snap701,1275040818,
 SNPC000709,heraNode700Snap709,1275040818,
+SNPA000700,heraNode700Snap0,1230372018,
+SNPB000701,heraNode700Snap1,1230372018
```
Comment (plaplant) [inline hera_mc/data/test_data/initialization_data_part_rosetta.csv:6]: I'm not familiar with the csv format. Do we need a trailing comma here, like in the previous lines?

File: hera_mc/mc_session.py [hunk 16]
```diff
@@ -1786,17 +1735,14 @@ def node_power_command(self, nodeID, part, command, nodeServerAddress=None,
...
         if nodeServerAddress is None:
-            nodeServerAddress = defaultServerAddress
+            nodeServerAddress = node.defaultServerAddress
 
         if testing:
             node_list = list(range(1, 31))
             dryrun = True
         else:
-            node_list = get_node_list(nodeServerAddress=nodeServerAddress)
+            node_list = node.get_node_list(nodeServerAddress=nodeServerAddress)
```
Comment (plaplant) [inline hera_mc/mc_session.py:1745]: It looks like this isn't covered by tests right now. If we're expecting this (because this is "live-only" functionality), then we should add a `# pragma: no cover` here.

File: hera_mc/mc_session.py [hunk 17]
```diff
@@ -9,16 +9,31 @@
...
 
+import yaml
+import numpy as np
 from sqlalchemy import desc, asc
 from sqlalchemy.orm import Session
 from sqlalchemy.sql.expression import func
 from astropy.time import Time, TimeDelta
 
 from .utils import get_iterable
+from .observations import Observation
```
Comment (plaplant) [inline hera_mc/mc_session.py:24]: Thanks for moving all of these here!

**Review by plaplant — COMMENTED** [review]
Thanks very much for the changes. I just have a few small comments, but I think this is nearly ready.

File: hera_mc/data/test_data/initialization_data_part_rosetta.csv (hunk 15 above)
Comment (bhazelton) [reply hera_mc/data/test_data/initialization_data_part_rosetta.csv:6]: Looks like no.

File: hera_mc/mc_session.py (hunk 16 above)
Comment (bhazelton) [reply hera_mc/mc_session.py:1745]: This is now covered by a test on the CI, thanks for catching it!

**Comment by bhazelton** [comment]
@plaplant I think I've addressed the comments and it's ready for another look.

**Review by plaplant — APPROVED** [review]
Thanks for all the great work @bhazelton and @david-deboer! I think this is good to go. I'll leave it up to someone else to merge because there are alembic issues, and I want to make sure we're ready to update the schema.
